File: go/0093M-复原IP地址.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-13 19:41:58
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func restoreIpAddresses(s string) []string {
	if len(s) < 4 || len(s) > 12 {
		return []string{}
	}
	check := func(start, end int) bool {
		if start > end || (s[start] == '0' && start != end) {
			return false
		}
		digit := 0
		for i := start; i <= end; i++ {
			if s[i] > '9' || s[i] < '0' {
				return false
			}
			digit = digit*10 + int(s[i]-'0')
			if digit > 255 {
				return false
			}
		}
		return true
	}
	ans := make([]string, 0)
	ip := ""
	var dfs func(index, cnt int)
	dfs = func(index, cnt int) {
		if cnt == 3 {
			if check(index, len(s)-1) {
				ans = append(ans, ip+s[index:])
			}
			return
		}
		for i := index; i < len(s); i++ {
			if check(index, i) {
				//l := len(ip)
				ip += string(s[index:i+1]) + "."
				dfs(i+1, cnt+1)
				ip = ip[:len(ip)-(i+2-index)]
			} else {
				return
			}
		}
	}
	dfs(0, 0)
	return ans
}

File: cpp/0093M-复原IP地址.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-27 14:00:06
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<string> ans;
    // string tmp;

    bool check(const string& s, int start, int end) {
        if (start == s.size()) return false;
        if (s[start] == '0' && start != end) return false;
        int t = 0;
        while (start <= end) {
            t = t * 10 + (s[start++] - '0');
            if (t > 255) return false;
        }
        return true;
    }

    void dfs(string s, int idx, int cnt) {
        if (cnt == 3) {
            if (check(s, idx, s.size() - 1)) ans.push_back(s);
            return;
        }
        for (int i = idx; i < s.size(); i++) {
            if (!check(s, idx, i)) break;
            s.insert(s.begin() + i + 1, '.');
            cnt++;
            dfs(s, i + 2, cnt);
            cnt--;
            s.erase(s.begin() + i + 1);
        }
    }

    vector<string> restoreIpAddresses(const string& s) {
        if (s.size() > 12) return ans;
        dfs(s, 0, 0);
        return ans;
    }
};